### DISCUSSION TITLE
add a 'double' package that distinguishes Stubs, Spies and Mocks

### DISCUSSION THREAD

**Discussion by laurentdutheil** [stretchr/testify#1643] [Ideas]
## Description
If we refer to a definition of [test doubles](https://martinfowler.com/bliki/TestDouble.html), there is an interest in distinguishing stubs, spies and mocks.

**Stubs** provide canned answers to calls made during the test. For example, a stub needs to return a value in response of a query. If the code under test changes and no longer needs to make that
query, there is no reason that the test should break.

**Spies** are stubs that also record some information based on how they were called. One form of this might be an email service that records how many messages it was sent.

**Mocks** are pre-programmed with expectations which form a specification of the calls they are expected to receive. They can throw an exception if they receive a call they don't expect and are
checked during verification to ensure they got all the calls they were expecting.

We can see the separation of concern between them as:

- a Stub handles the predefined calls
- a Spy is a Stub who records the actual calls.
- a Mock is a Spy who can make assertions and fail the tests on error.

```mermaid
classDiagram
    Stub <|-- Spy
    Spy <|-- Mock

    class Stub {
        - predefinedCalls
        + Called()
        + On()
        ...()
    }
    class Spy {
        - actualCalls
        + NumberOfCalls()
        ...()
    }
    class Mock {
        + AssertCalled()
        ...()
    }

```

## Proposed solution
There is a POC [here](https://github.com/laurentdutheil/go-double)

I try to explain the design choices in [Architecture Decision Records](https://github.com/laurentdutheil/go-double/tree/main/adr).
All the ADR are negotiable, except perhaps the first one which is the purpose of the POC.

Opinion:
- a Stub should fail the test if we don't know what to return
- a Stub should not fail the test if a prepared call is not called
- a Stub should not fail the test if a method without return argument was not in the prepared calls

The POC introduces reflection (to know if a method have return parameters) and generics (to propose a constructor for Stubs, Spies and Mocks)

The purpose is to be as retro-compatible as possible to facilitate the migration.

It is opinionated. So I'll understand a reject. Even so, it was fun to code.

## Use case
There is a [comparison](https://github.com/laurentdutheil/go-double/blob/main/comparaison_test.go) between the actual library and the proposition.

There is also examples of [Stub](https://github.com/laurentdutheil/go-double/blob/main/stub_example_test.go), [Spy](https://github.com/laurentdutheil/go-double/blob/main/spy_examples_test.go) and [Mock](https://github.com/laurentdutheil/go-double/blob/main/mock_example_test.go)

## References
Can be referred to #483

**Comment by brackendawson** [comment]
When I talk in this area I use the language from Sandi Metz's book ([poodr.com](https://poodr.com), chapter 9) which uses only "doubles" and "mocks".
* Doubles: Reply to calls with information. Whether they are called or not, or how many times, is not checked.
* Mocks: Are calls which are ecpected to have a side-effect, and so we must check that the call is made.

While "testing for state" as Sandi recommends, I've never had the need for a "spy". This shouldn't be relevant to what I'm about to say, though.

New packages or classes in this fork of testify are, in general, highly unlikely at this point. I think the current API is sufficient to do both and this is how I contstruct them by using variale naming and properly setting up the `mock.Mock` class.

## Doubles
```go
databaseDouble := &myMock{}
databaseDouble.Test(t) // Prevents panic on unexpected calls
databaseDouble.On("GetFullName", "mat").Return("Mat Ryer", nil)
// ...and so on
```
This will return the information you tell it to. Changing your implementation is unlikely to need a change to the double.

## Mocks
```go
mockDatabase := &myMock{}
mockDatabase.Test(t)
defer mockDatabase.AssertExpectations(t) // All calls were made or fail the test
mockDatabase.On("Delete", "mat").Return(nil).Once() // All calls should have an expected count
// ...and so on
```
This will fail the test if the code under test didn't try to delete Mat.

## What about both
You should consider whether your implementation is optimal before you do this, but if you can't or don't want to avoid your duck being a mock and double, you can do this:
```go
mockDatabase := &myMock{}
mockDatabase.Test(t)
defer mockDatabase.AssertExpectations(t)
mockDatabase.On("GetFullName", "mat").Return("Mat Ryer", nil).Maybe() // This call is a double
mockDatabase.On("Delete", "mat").Return(nil).Once()
```
Here `GetFullName` can be called or not as many times as you want, but `Delete` must be called once.

## Spies
I don't use them, but I expect you could use `mock.Mock.Calls` to create one.

**Comment by laurentdutheil** [comment]
Thank you for your feedback.

I'm a fan of the work of Sandi Metz. I particularly like this [talk](https://www.youtube.com/watch?v=URSWYvyc42M).
My other reference is this article of Martin Fowler: [Mocks Aren't Stubs](https://martinfowler.com/articles/mocksArentStubs.html)

I agree with you that it is possible to write stubs and mocks with the current version.

However, I find some drawbacks:
- When a method have no return value, I have to specify the call in my test. Without this, the "mock" fail as it don't know what to return.
`mockDatabase.On("DoSomethingWith", "this").Maybe()`
- As I understand [here](https://github.com/stretchr/testify/issues/558), the `Unset` method exists because of this limitation. When people want to share a Stub/Mock definition between several tests, they have to unset some behavior depending on what they want to test.

I find some advantages to separate Stub and Mock notions:
- For people like me who prefer to make the distinction, the tests are easier to read. When I use a Stub, I make assertion on the state. When I use Mock, I make assertion on the delegation. (as in the Sandi Metz talk)
- For the maintainability of the library, the distinction permits to separate the concerns. And perhaps, we don't need `Maybe` or `Unset` methods anymore

Do you think it is worth distinguishing stubs and mocks?

**Reply by brackendawson** [reply]
In my opinion `mock.Call.Unset` should not have been added to the testify API, and mocks should never be shared between tests. People should set up a new mock for each test with the expectations up front and then test for state. *If you know you need to remove an expectation, you can also know you don't need to add it in the first place.* However it was added to our API and now we have to support it.

Removing the need for priming each call would be a benefit, creating a "blind mock", but only works for calls with no returns at all. I feel like this is pretty niche as a lot of calls with only a side-effect will still return an error. Perhaps adding the ability to pass `mock.Anything` as the function name would help here?

A larger concern about adapting the mock package around one definition of types of mocks is that there are many different definitions of mocks, many of them perfectly good. By throwing in behind Martin's do we make it harder for people to follow Sandi's?

If the outcome of this change would be to remove methods like Maybe and Unset then this will definitely not happen. Testify is being maintained as v1.
